File: src/data/hullmods/niko_MPC_superPDComputer.java
package data.hullmods;

import com.fs.starfarer.api.combat.BaseHullMod;
import com.fs.starfarer.api.combat.MutableShipStatsAPI;
import com.fs.starfarer.api.combat.ShipAPI;
import com.fs.starfarer.api.impl.campaign.ids.Stats;

public class niko_MPC_superPDComputer extends BaseHullMod {

    public final float missileDamagePercent = 50f;
    public final float rangePercent = 5000f;

    @Override
    public void applyEffectsBeforeShipCreation(ShipAPI.HullSize hullSize, MutableShipStatsAPI stats, String id) {

        stats.getNonBeamPDWeaponRangeBonus().modifyPercent(id, rangePercent);
        stats.getBeamPDWeaponRangeBonus().modifyPercent(id, rangePercent);

        stats.getDynamic().getMod(Stats.PD_IGNORES_FLARES).modifyFlat(id, 1f);
        stats.getDynamic().getMod(Stats.PD_BEST_TARGET_LEADING).modifyFlat(id, 1f);
        stats.getDamageToMissiles().modifyPercent(id, missileDamagePercent);
    }

    public String getDescriptionParam(int index, ShipAPI.HullSize hullSize) {
        if (index == 1) return "" + (int)Math.round(rangePercent) + "%";
        else if (index == 0) return "" + (int)Math.round(missileDamagePercent) + "%";
        return null;
    }

}

File: src/data/hullmods/niko_MPC_superPDComputer.kt
package data.hullmods

import com.fs.starfarer.api.combat.BaseHullMod
import com.fs.starfarer.api.combat.MissileAPI
import com.fs.starfarer.api.combat.MutableShipStatsAPI
import com.fs.starfarer.api.combat.ShipAPI
import com.fs.starfarer.api.combat.ShipAPI.HullSize
import com.fs.starfarer.api.combat.WeaponAPI
import com.fs.starfarer.api.combat.WeaponAPI.AIHints
import com.fs.starfarer.api.combat.WeaponAPI.WeaponType
import com.fs.starfarer.api.impl.campaign.ids.Stats
import com.fs.starfarer.api.loading.ProjectileSpawnType
import com.fs.starfarer.api.loading.ProjectileSpecAPI
import data.utilities.niko_MPC_battleUtils.isPD

class niko_MPC_superPDComputer : BaseHullMod() {

    override fun applyEffectsAfterShipCreation(ship: ShipAPI?, id: String?) {
        super.applyEffectsAfterShipCreation(ship, id)

        val weapons: List<WeaponAPI> = ship!!.allWeapons
        val iter = weapons.iterator()
        while (iter.hasNext()) {
            val weapon = iter.next()
            if (weapon.isPD()) continue
            if (weapon.size == WeaponAPI.WeaponSize.SMALL) {
                val projSpec: ProjectileSpecAPI? = weapon.spec.projectileSpec as? ProjectileSpecAPI
                if (projSpec != null && weapon.type != WeaponType.MISSILE && !weapon.hasAIHint(AIHints.STRIKE)){
                    weapon.setPD(true)
                    return
                }
            }
            weapon.setPDAlso(true) //CHAOS
        }
    }
}